test/controllers/admin/deploy_groups_controller_test.rb

```diff
@@ -232,5 +232,99 @@
+        end
+
+        it "removes the next_stage_id" do
+          assert template_stage.reload.next_stage_ids.include?(stage.id)
+
+          post :merge_all_stages, params: {id: deploy_group}
+
+          refute template_stage.reload.next_stage_ids.include?(stage.id)
+        end
+
+        it "adds the deploy group to the template stage" do
+          refute template_stage.deploy_groups.include?(deploy_group)
+
+          post :merge_all_stages, params: {id: deploy_group}
+
+          assert template_stage.deploy_groups.include?(deploy_group)
+        end
+
+        it "removes the stale deploy_group links as well" do
+          assert DeployGroupsStage.where(stage_id: stage.id, deploy_group_id: deploy_group.id).first
+
+          post :merge_all_stages, params: {id: deploy_group}
+
+          refute DeployGroupsStage.where(stage_id: stage.id, deploy_group_id: deploy_group.id).first
+        end
+
+        it "replaces the stale stage deploy group link with one for the template stage" do
+          refute DeployGroupsStage.where(stage_id: template_stage.id, deploy_group_id: deploy_group.id).first
+
+          post :merge_all_stages, params: {id: deploy_group}
+
+          assert DeployGroupsStage.where(stage_id: template_stage.id, deploy_group_id: deploy_group.id).first
+        end
+      end
+    end
+
+    describe "with multiple stages to remove" do
+
+      let(:env) {environments(:staging)}
+      let(:template_deploy_group) {deploy_groups(:pod100)} # needed so the templaste stages have an environment.
+      let(:deploy_group) {DeployGroup.create!(name: 'Pod 101', environment: env)}
+
+      before do
+        # create 2 new template stages, and remember we still have a 3rd from the default fixtures.
+
+        project1 = Project.create!(name: "foo1", include_new_deploy_groups: true, permalink: "foo1", repository_url: "https://github.com/samson-test-org/example-project.git")
+        Stage.create!(name: "foo1 tstage", project: project1, is_template: true, deploy_groups: [template_deploy_group])
+
+        project2 = Project.create!(name: "foo2", include_new_deploy_groups: true, permalink: "foo2", repository_url: "https://github.com/samson-test-org/example-project.git")
+        Stage.create!(name: "foo2 tstage", project: project2, is_template: true, deploy_groups: [template_deploy_group])
+
+        # now create the non-template stages for this deploy-group
+        Admin::DeployGroupsController.create_all_stages(deploy_group)
+      end
+
+      it "merges and soft-deletes all non-template stages" do
+        assert_equal 3, deploy_group.stages.count
+
+        post :merge_all_stages, params: {id: deploy_group}
+
+        # should have only template-stages now (all others were soft-deleted)
+        stages = deploy_group.stages.where(is_template: false)
+        Rails.logger.warn("stages after soft delete #{ stages.map { |s| [s.reload.project.name, s.name, s.deleted?] }.inspect }")
+        assert_empty stages
+      end
```

Comment:
@grosser I've isolated the soft deletion issue with this test. `soft-delete` is definitely called on all 3 stages. Only one stage ever gets soft-delete, all other calls to soft-delete! become NOOP.

This is what the log looks like when soft-delete is called. The last line is my rails log line to verify it was soft deleted properly.

```
   (0.4ms)  SAVEPOINT active_record_1
   (0.5ms)  RELEASE SAVEPOINT active_record_1
  Stage Load (0.8ms)  SELECT  `stages`.* FROM `stages` WHERE `stages`.`id` = 685640165 LIMIT 1
Soft delete of stage 685640165 failed
```